--- dwm.c.orig  2013-11-29 13:58:49.673588564 +0100
+++ dwm.c   2013-11-23 22:56:49.725007370 +0100
@@ -193,6 +193,7 @@                                                                                                            
 static void resizemouse(const Arg *arg);                                                                                      
 static void restack(Monitor *m);                                                                                              
 static void run(void);                                                                                                        
+static void runorraise(const Arg *arg);
 static void scan(void);                                                                                                       
 static Bool sendevent(Client *c, Atom proto);                                                                                 
 static void sendmon(Client *c, Monitor *m);                                                                                   
@@ -1349,6 +1350,30 @@                                                                                                         
 }                                                                                                                             
                                                                                                                               
 void                                                                                                                          
+runorraise(const Arg *arg) {
+    char *app = ((char **)arg->v)[4];
+    Arg a = { .ui = ~0 };
+    Monitor *mon;
+    Client *c;
+    XClassHint hint = { NULL, NULL };
+    /* Tries to find the client */
+    for (mon = mons; mon; mon = mon->next) {
+        for (c = mon->clients; c; c = c->next) {
+            XGetClassHint(dpy, c->win, &hint);
+            if (hint.res_class && strcmp(app, hint.res_class) == 0) {
+                a.ui = c->tags;
+                view(&a);
+                focus(c);
+                XRaiseWindow(dpy, c->win);
+                return;
+            }
+        }
+    }
+    /* Client not found: spawn it */
+    spawn(arg);
+}
+
+void
 scan(void) {                                                                                                                  
    unsigned int i, num;                                                                                                       
}
