diff --git a/config.def.h b/config.def.h
index 77ff358..203d354 100644
--- a/config.def.h
+++ b/config.def.h
@@ -12,6 +12,7 @@ static const unsigned int borderpx  = 1;        /* border pixel of windows */
 static const unsigned int snap      = 32;       /* snap pixel */
 static const Bool showbar           = True;     /* False means no bar */
 static const Bool topbar            = True;     /* False means bottom bar */
+static const unsigned int bottommargin = 20;    /* Margin at the bottom for another bar */
 
 /* tagging */
 static const char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };
diff --git a/dwm.c b/dwm.c
index 1d78655..cece290 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1841,6 +1841,9 @@ updatebarpos(Monitor *m) {
 	m->wh = m->mh;
 	if(m->showbar) {
 		m->wh -= bh;
+        m->wh -= marginbottom;
+        if(!m->topbar)
+            m->wy += marginbottom;
 		m->by = m->topbar ? m->wy : m->wy + m->wh;
 		m->wy = m->topbar ? m->wy + bh : m->wy;
 	}
